lgc/patch/PatchBoundsCheckMemory.cpp

```diff
@@ -0,0 +1,195 @@
+// Executes this pass on the specified LLVM module.
+//
+// @param [in/out] module : LLVM module to be run on
+bool PatchBoundsCheckMemory::runOnModule(Module &module) {
+  LLVM_DEBUG(dbgs() << "Run the pass llpc-bounds-check-memory\n");
+
+  m_pipelineState = getAnalysis<PipelineStateWrapper>().getPipelineState(&module);
+
+  if (!m_pipelineState->getOptions().enableScratchBoundsCheck)
+    return false;
+
+  m_context = &module.getContext();
+  m_builder.reset(new IRBuilder<>(*m_context));
+
+  visit(module);
+
+  bool changed = false;
+  for (auto &inst : m_getElementPtrInsts)
+    changed |= addBoundsCheck(inst);
+  m_getElementPtrInsts.clear();
+
+  return changed;
+}
+
+// =====================================================================================================================
+// Visits "getelementptr" instruction.
+//
+// @param getElemPtr : "GetElementPtr" instruction
+void PatchBoundsCheckMemory::visitGetElementPtrInst(GetElementPtrInst &getElemPtr) {
+  auto pointerTy = getElemPtr.getPointerOperand()->getType();
+  if (pointerTy->getPointerAddressSpace() != ADDR_SPACE_PRIVATE)
+    return;
+
+  // Check for dynamic indices in the instruction
+  llvm::SmallVector<std::pair<Value *, uint64_t>, 1> dynIndices;
+  std::vector<Value *> indices;
+  for (unsigned i = 1, operandCount = getElemPtr.getNumOperands(); i < operandCount; ++i) {
+    auto index = getElemPtr.getOperand(i);
+    if (!isa<Constant>(index)) {
+      auto indexedTy = GetElementPtrInst::getIndexedType(pointerTy->getPointerElementType(), indices);
+      if (auto vectorTy = dyn_cast<FixedVectorType>(indexedTy)) {
+        dynIndices.push_back({index, vectorTy->getNumElements()});
+      } else if (auto arrayTy = dyn_cast<ArrayType>(indexedTy)) {
+        dynIndices.push_back({index, arrayTy->getNumElements()});
+      } else {
+        llvm_unreachable("Unsupported indexed type for bounds checking");
+      }
+      break;
+    }
+    indices.push_back(index);
+  }
+  if (dynIndices.empty())
+    return;
+  m_getElementPtrInsts.push_back({&getElemPtr, dynIndices});
+}
+
+// =====================================================================================================================
+// Checks whether each dynamic index in the specified "getelementptr" instruction is lower than its allowed upper bound.
+//
+// @param info : "GetElementPtr" instruction and its dynamic indinces
+// @returns : True of the code was changed or false if unmodified
+bool PatchBoundsCheckMemory::addBoundsCheck(GetElemPtrInfo &info) {
+  auto &getElemPtr = *info.getElemPtr;
+
+  // Copy users
+  std::vector<User *> users;
+  for (auto user : getElemPtr.users()) {
+    // Look through bitcasts
+    Instruction *inst = cast<Instruction>(user);
+    if (isa<BitCastInst>(inst)) {
+      for (auto user : inst->users())
+        users.push_back(user);
+    } else {
+      users.push_back(user);
+    }
```

Comment:
Can you use `inst->stripPointerCasts()` here?

Reply:
I want to look through bitcasts of the result of getelementptr/the users. `stripPointerCasts` does look at the used inputs to the instruction, which is the wrong direction as far as I understand.